content/lib/2021-04-02-operating-system-interface.md

```diff
@@ -0,0 +1,225 @@
+
+ - CSI defines the container storage API surface/shape for interfacing with block and file storage drivers in `userland`. 
+     - More information on [CSI drivers](https://kubernetes.io/blog/2019/01/15/container-storage-interface-ga/)
+     - More information on [Kubelet device registration and plugins](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/)
+ - CSI does not solve *how* to install or manage block or file storage drivers.
+ - The Kubelet plugin system does NOT solve how to install or manage Kubelet plugins from Kubernetes. 
+     - [Access to the Kubelet API server is required](https://v1-18.docs.kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-registration) to interact with the exposed registration service. 
+ - CSI (via the Kubelet registration endpoint and logic) can set limitations and provide status updates in `kubernetesland`.
+
+
+### Runtime (Compute) (CRI/OCI)
+
+Official [Container Runtime Interface](https://github.com/kubernetes/cri-api) API.
+Official [Open Container Interface](https://github.com/opencontainers/runtime-spec) standard.
+    - Official [Open Container Image Spec](https://github.com/opencontainers/image-spec)
+
+CRI is a implementation of OCI.
+The main difference is scope, in which case OCI also defines the [open container image spec](https://github.com/opencontainers/image-spec).
+
+
+- OCI and CRI define the container runtime API surface/shape for interfacing with container runtime "engines" from `userland`. 
+    - OCI also defines the API surface/shape and interface for packaging container images
+- Container runtimes are tightly coupled with Linux features, and therefor are tightly coupled with the operating system. 
+- Container runtimes manage permissions, storage, and networking in different ways. 
+- Container runtimes are tightly coupled with `storage` and `networking`.
+- Container runtimes do not define *how* they are installed or managed. 
+
+# The Void
+
+There are several `userland` components that are not considered from a management perspective. 
+
+Below are demonstrable questions that technically have answers.
+I hope to challenge the existing answers to these questions with DOSi.
+
+### Demonstrable Questions
+
+- How is my container network defined?
+- How are my storage drivers defined?
+- How is my container runtime defined?
+    - Which container runtime will I use?
+    - How will I upgrade my container runtime?
+- Which clients have access to my container runtime?
+- How do I mutate my network configuration in `userland`?
+- How do I mutate my network configuration in `kubernetesland`?
+- How do I mutate kernel configuration? 
+    - eBPF
+    - Kernel modules
+    - IPTables
+    - Seccomp filters and Seccomp policy
+    - Namespace sharing
+    - Linux capabilities
+- How can I read kernel runtime values?
+
```

Comment:
Some other things we feel are important in Talos:

- How do I manage OS upgrades in the context of distributed systems? (i.e. How do ensure I don't kill `etcd`, or imagine being able to have the OS's upgrade policy be aware of CSI, and wait for replication before upgrading.)
- How should I secure my node in the context of distributed systems? Talos, for example, is read only and runs in RAM. Kubernetes writes to disk. This has proven to be enough if all we wan't is something Kubernetes-like.